samples/snippets/storage_transfer_manager.py

```diff
@@ -0,0 +1,136 @@
+
+
+def upload_many_blobs_with_transfer_manager(bucket_name, filenames, root=""):
+    """Upload every file in a list to a bucket, concurrently in a thread pool.
+
+    Each blob name is derived from the filename, not including the `root`
+    parameter. For complete control of the blob name for each file (and other
+    aspects of individual blob metadata), use transfer_manager.upload_many()
+    instead.
+    """
+
+    # The ID of your GCS bucket
+    # bucket_name = "your-bucket-name"
+
+    # A list (or other iterable) of filenames to upload.
+    # filenames = ["file_1.txt", "file_2.txt"]
+
+    # The path on your computer that is the root of all of the files in the list
+    # of filenames. This string is prepended to the each filename to get the
+    # full path to the file. Be sure to include the trailing slash. Relative
+    # paths and absolute paths are both accepted. This string is *not* included
+    # in the name of the uploaded blob; it is only used to find the source
+    # files. An empty string means "the current working directory".
+    # root=""
+
+    storage_client = Client()
+    bucket = storage_client.bucket(bucket_name)
+
+    results = transfer_manager.upload_many_from_filenames(bucket, filenames, root=root)
+
+    for name, result in zip(filenames, results):
+        # The results list is either `None` or an exception for each filename in
+        # the input list, in order.
+
+        if isinstance(result, Exception):
+            print("Failed to upload {} due to exception: {}".format(name, result))
+        else:
+            print("Uploaded {} to {}.".format(name, bucket.name))
+
+
+def download_blob_chunks_concurrently_with_transfer_manager(
+    bucket_name, blob_name, local_filename, chunk_size=200 * 1024 * 1024
+):
+    """Download a single blob, in chunks, concurrently in a thread pool.
+
+    This is intended for use with very large blobs."""
+
+    # The ID of your GCS bucket
+    # bucket_name = "your-bucket-name"
+
+    # The name of the blob to download.
+    # blob_name = "your-blob.txt"
+
+    # The filename (or path) on your computer to which to download the blob.
+    # local_filename = "your-file.txt"
+
+    # The size of each chunk. The file will be divided into as many pieces as
+    # needed based on this chunk size. For instance, if the chunk size is
+    # 200 megabytes, a 1.6 gigabyte file would be downloaded in eight pieces
+    # concurrently.
+    # chunk_size = 200 * 1024 * 1024  # 200 MiB.
+
+    storage_client = Client()
+    bucket = storage_client.bucket(bucket_name)
+
+    blob = bucket.blob(blob_name)
+
+    # Open the local file in binary write mode.
+    local_file = open(local_filename, "wb")
+
+    # Unlike other transfer manager functions, which handle multiple files and
+    # return exceptions in a list, this function will simply raise any exception
+    # it encounters, and has no return value.
+    transfer_manager.download_chunks_concurrently_to_file(
```

Comment:
Thoughts on call this download_ file_obj  or something simpler? It's basically a "smart" version download that handles the range reads for you.

Reply:
I struggled with that nomenclature for sure, and ended up on this which is quite verbose. I'd be happy to change it. My main goals is differentiation between it and other download methods, and consistency with the existing "to_file" or "to_filename" naming pattern. Over the years I've seen many types of software that try to do multithreaded downloads and they're all named awkwardly (like this) or by weird brand names.

"download_to_file_threaded" or "download_chunks_to_file" could work if we wanted to slim it down a touch.